h/static/scripts/annotation-sync.coffee

```diff
@@ -0,0 +1,190 @@
+class AnnotationSync
+  # Default configuration
+  options:
+    # Formats an annotation into a message body for sending across the bridge.
+    formatter: (annotation) -> annotation
+
+    # Recieves an annotation extracted from the message body received
+    # via the bridge and returns an annotation for use in the local app.
+    parser: (annotation) -> annotation
+
+    # Merge function. If specified, it will be called with the local copy of
+    # an annotation and a parsed copy received as an argument to an RPC call
+    # to reconcile any differences. The default behavior is to merge all
+    # keys of the remote object into the local copy
+    merge: (local, remote) ->
+      for k, v of remote
+        local[k] = v
+      local
+
+    # Function used to emit annotation events
+    emit: (event, args...) ->
+      throw new Error('options.emit unspecified for AnnotationSync.')
+
+    # Function used to register handlers for annotation events
+    on: (event, handler) ->
+      throw new Error('options.on unspecified for AnnotationSync.')
+
+  # Cache of annotations which have crossed the bridge for fast, encapsulated
+  # association of annotations received in arguments to window-local copies.
+  cache: null
+
+  constructor: (options, @bridge) ->
+    @options = $.extend(true, {}, @options, options)
+
+    @cache = {}
+
+    @_on = @options.on
+    @_emit = @options.emit
+
+    # Listen locally for interesting events
+    for event, handler of @_eventListeners
+      this._on(event, handler.bind(this))
+
+    # Register remotely invokable methods
+    for method, func of @_channelListeners
+      @bridge.on(method, func.bind(this))
+
+    # Upon new connections, send over the items in our cache
+    onConnect = (channel) =>
+      this._syncCache(channel)
+    @bridge.onConnect(onConnect)
+
+  # Provide a public interface to the annotation cache so that other
+  # sync services can lookup annotations by tag.
+  getAnnotationForTag: (tag) ->
+    @cache[tag] or null
+
+  sync: (annotations, cb) ->
+    annotations = (this._format a for a in annotations)
+    @bridge.call
+      method: 'sync'
+      params: annotations
+      callback: cb
+    this
+
+  # Handlers for messages arriving through a channel
+  _channelListeners:
+    'beforeCreateAnnotation': (txn, body) ->
```

Comment:
Indeed. There's this and a few more things that I would like to refactor, but first let's get everything working.